SimPEG/EM/NSEM/ProblemNSEM.py

```diff
@@ -0,0 +1,599 @@
+                dRHS_dm_v = self.getRHSDeriv(freq, v) # Size: nE,2 (u_px,u_py) in the columns.
+                # Calculate du/dm*v
+                du_dm_v = Ainv * ( - dA_dm_v + dRHS_dm_v)
+                # Calculate the projection derivatives
+                for rx in src.rxList:
+                    # Calculate dP/du*du/dm*v
+                    Jv[src, rx] = rx.evalDeriv(src, self.mesh, f, mkvc(du_dm_v)) # wrt uPDeriv_u(mkvc(du_dm))
+            Ainv.clean()
+        # Return the vectorized sensitivities
+        return mkvc(Jv)
+
+    def Jtvec(self, m, v, f=None):
+        """
+        Function to calculate the transpose of the data sensitivities (dD/dm)^T times a vector.
+
+            :param numpy.ndarray m (nP,) - inversion model
+            :param numpy.ndarray v (nD,) - vector which we take adjoint product with (nP,)
+            :param NSEMfields object f (optional) - NSEM fields object, if not given it is calculated
+            :rtype: numpy.array:
+            :return: Jtv (nP,) Data sensitivities wrt m
+        """
+
+        if f is None:
+            f = self.fields(m)
+
+        self.curModel = m
+
+        # Ensure v is a data object.
+        if not isinstance(v, self.dataPair):
+            v = self.dataPair(self.survey, v)
+
+        Jtv = np.zeros(m.size)
+
+        for freq in self.survey.freqs:
+            AT = self.getA(freq).T
+
+            ATinv = self.Solver(AT, **self.solverOpts)
+
+            for src in self.survey.getSrcByFreq(freq):
+                # u_src needs to have both polarizations
+                u_src = f[src, :]
+
+                for rx in src.rxList:
+                    # Get the adjoint evalDeriv
+                    # PTv needs to be nE,2
+                    PTv = rx.evalDeriv(src, self.mesh, f, mkvc(v[src, rx]), adjoint=True) # wrt f, need possibility wrt m
+                    # Get the
+                    dA_duIT = mkvc(ATinv * PTv) # Force (nU,) shape
+                    dA_dmT = self.getADeriv(freq, u_src, dA_duIT, adjoint=True)
+                    dRHS_dmT = self.getRHSDeriv(freq, dA_duIT, adjoint=True)
+                    # Make du_dmT
+                    du_dmT = -dA_dmT + dRHS_dmT
+                    # Select the correct component
+                    # du_dmT needs to be of size (nP,) number of model parameters
+                    real_or_imag = rx.component
+                    if real_or_imag == 'real':
+                        Jtv +=  np.array(du_dmT,dtype=complex).real
+                    elif real_or_imag == 'imag':
+                        Jtv +=  -np.array(du_dmT,dtype=complex).real
+                    else:
+                        raise Exception('Must be real or imag')
+            # Clean the factorization, clear memory.
+            ATinv.clean()
+        return Jtv
+
+###################################
+## 1D problems
+###################################
+
+class Problem1D_ePrimSec(BaseNSEMProblem):
```

Comment:
The primary/secondary more refers to the `Source` of the problem. But since in other formulation's some of the `Problem` methods will be different, such as `getA` and `getRHS` this additional distinction was made. 
I agree with find some of the naming confusing but I reluctant to address these issue in this PR. The naming convention has to be considered over the whole `NSEM` since `Sources`, `Fields` and `Problem` are tied together depending on formulation.